configmap/informed_watcher.go

```diff
@@ -61,24 +62,61 @@ type InformedWatcher struct {
 	informer corev1informers.ConfigMapInformer
 	started  bool
 
+	// cfgs are the default ConfigMaps to use if the real ones do not exist or are deleted.
+	cfgs map[string]*corev1.ConfigMap
+
 	// Embedding this struct allows us to reuse the logic
 	// of registering and notifying observers. This simplifies the
-	// InformedWatcher to just setting up the Kubernetes informer
+	// InformedWatcher to just setting up the Kubernetes informer.
 	ManualWatcher
 }
 
 // Asserts that InformedWatcher implements Watcher.
 var _ Watcher = (*InformedWatcher)(nil)
 
-// Start implements Watcher
+// Asserts that InformedWatcher implements DefaultingWatcher.
+var _ DefaultingWatcher = (*InformedWatcher)(nil)
+
+// WatchWithDefault implements DefaultingWatcher.
+func (i *InformedWatcher) WatchWithDefault(cm corev1.ConfigMap, o Observer) {
+	if i.cfgs == nil {
+		i.cfgs = map[string]*corev1.ConfigMap{}
+	}
+	i.cfgs[cm.Name] = &cm
+
+	i.m.Lock()
+	started := i.started
+	i.m.Unlock()
+	if started {
+		// TODO make both Watch and WatchWithDefault work after the InformedWatcher has started.
+		// This likely entails changing this to `o(&cm)` and having Watch check started, if it has
+		// started, then ensuring i.informer.Lister().ConfigMaps(i.Namespace).Get(cmName) exists and
+		// calling this observer on it. It may require changing Watch and WatchWithDefault to return
+		// an error.
+		panic(errors.New("cannot WatchWithDefault after the InformedWatcher has started"))
+	}
+
+	i.Watch(cm.Name, o)
+}
+
+// Start implements Watcher.
 func (i *InformedWatcher) Start(stopCh <-chan struct{}) error {
+	// Pretend that all the defaulted ConfigMaps were just created. This is done before we start
+	// the informer to ensure that if a defaulted ConfigMap does exist, then the real value is
+	// processed after the default one.
+	for k := range i.observers {
+		if cfg, ok := i.cfgs[k]; ok {
+			i.addConfigMapEvent(cfg)
+		}
+	}
+
 	if err := i.registerCallbackAndStartInformer(stopCh); err != nil {
 		return err
 	}
 
 	// Wait until it has been synced (WITHOUT holing the mutex, so callbacks happen)
 	if ok := cache.WaitForCacheSync(stopCh, i.informer.Informer().HasSynced); !ok {
-		return errors.New("Error waiting for ConfigMap informer to sync.")
+		return errors.New("error waiting for ConfigMap informer to sync")
 	}
 
 	return i.checkObservedResourcesExist()
```

Comment:
The point of this is to block until we've seen each thing at least once.  I am curious in which scenarios that might be inadequate?

Reply:
This doesn't block, it returns an error if a ConfigMap doesn't. 

As for why, from #412:

This has come up for dynamic tracing support in knative/eventing. We want to allow per namespace tracing, so we can't simply precreate the config-tracing ConfigMap while installing eventing. 

We expect the ConfigMap to be absent most of them time. Only when the user has decided they want tracing in this namespace, will they create it. At which point, we want the new config picked up without requiring a restart of all the Pod.

Our expectation is that we have tracing either disabled or sampling set to zero in a code default. Have each Pod we control run the configmap.InformedWatcher so that when the ConfigMap is created/updated/deleted the Pod immediately notices and starts using the updated config.